core/src/main/scala/com/github/mrpowers/spark/fast/tests/DatasetComparer.scala

```diff
@@ -66,12 +66,12 @@ Expected DataFrame Row Count: '$expectedCount'
       assertSmallDatasetContentEquality(defaultSortDataset(actualDS), defaultSortDataset(expectedDS), truncate, equals)
   }
 
-  def assertSmallDatasetContentEquality[T](actualDS: Dataset[T], expectedDS: Dataset[T], truncate: Int, equals: (T, T) => Boolean): Unit = {
+  def assertSmallDatasetContentEquality[T: ClassTag](actualDS: Dataset[T], expectedDS: Dataset[T], truncate: Int, equals: (T, T) => Boolean): Unit = {
     val a = actualDS.collect().toSeq
     val e = expectedDS.collect().toSeq
     if (!a.approximateSameElements(e, equals)) {
       val arr = ("Actual Content", "Expected Content")
```

Comment:
I don`t see the improvement, any Dataset[T] can be transformed into a DataFrame, the only change here is that it will work with the original Class `T` that was provided, but the result will be the same, only with the Class name prefixing each line. If the only thing was to prevent the .asRows method we could have done this and not any other change will be needed.
```suggestion
    val a = actualDS.toDF.collect() //now its a Row
    val e = expectedDS.toDF.collect()
    if (!a.approximateSameElements(e, equals)) {
      val arr = ("Actual Content", "Expected Content")
      val msg = "Diffs\n" ++ DataframeUtil.showDataframeDiff(arr, a, e, truncate)
```

I see this as more unified and we wont have to check if its a Product or not.

Reply:
`ShowProductDiff` is an update/extension to the original `showDataframeDiff` so we can have a nicely format table display for any `Case Class` and additionally Row. This was done so we have nicely format table for `StructField`

And since `showProductDiff` can now also display differences for any `Case Class` I figured we no longer need to convert a `Dataset[T]` into a `Dataframe`.

Reply:
since we also use it for Seq[StructField] for displaying schema fields won't we have to make actual and expected `Seq[Any]` ? 

```scala
private[mrpowers] def showDataframeDiff(
      header: (String, String),
      actual: Seq[Any],
      expected: Seq[Any],
      truncate: Int = 20,
      minColWidth: Int = 3,
     className = Option[String]
  ): String = {
   val (className, lBracket, rBracket)  = className match { 
     case None => ("", "[", "]") 
     case Some(cn) (cn, "(", ")")
    }

```

Reply:
```scala
def assertSmallDatasetContentEquality[T: ClassTag](actualDS: Dataset[T], expectedDS: Dataset[T], truncate: Int, equals: (T, T) => Boolean): Unit = {
    val a = actualDS.collect().toSeq
    val e = expectedDS.collect().toSeq
    if (!a.approximateSameElements(e, equals)) {
      val arr = ("Actual Content", "Expected Content")
      val runTimeClass                     = implicitly[ClassTag[T]].runtimeClass
      val prefix = if (runTimeClass == classOf[Row]) None else Some(runTimeClass.getSimpleName)
      val msg = "Diffs\n" ++ DataframeUtil.showDataframeDiff(arr, a.asRows, e.asRows, truncate, prefix)
      throw DatasetContentMismatch(msg)
    }
  }
```

If you don't want to do this here, move it inside `showDataframeDiff`, you can put the type parameter but you ar not forced to use it. But I see it weird to have the signature of your function as:

```scala
private[mrpowers] def showDataframeDiff[T:ClassTag](
      header: (String, String),
      actual: Seq[Any],
      expected: Seq[Any],
      truncate: Int = 20,
      minColWidth: Int = 3
  ): String
```

Reply:
I prefer putting it in `showProductDiff` since it will also need to be on `betterSchemaMismatchMessage`. Do you think we should still use `Seq[T]`? I think that would make it at least safer for the input Type.

Reply:
ok to change the name, it's called from a dataset. To keep the `[T]`, this is not a public method, it carries the type and the compiler is going to validate the schema of the actual and expected dataframe. The only problem is if the conversion to DataSet[T] is not correct, e.g. :
```scala
assertSmallDataset(spark.table("aTable").as[MyType], expected)
```
If the table doesn't match, the error will be raised with an exception in the `.as[T]` method in runtime.
That is the reason I say if we detect a non-Row DataSet, to skip the validation of the schema because it's already done.

Reply:
@alfonsorr I fixed issue with Dataset[Array[Int]] and also add test cases for String, Int, Array, Seq cases